pkg/reconciler/v1alpha1/configuration/configuration.go

```diff
@@ -230,3 +253,61 @@ func (c *Reconciler) updateStatus(u *v1alpha1.Configuration) (*v1alpha1.Configur
 	}
 	return newu, nil
 }
+
+func (c *Reconciler) gcRevisions(ctx context.Context, config *v1alpha1.Configuration) error {
+	logger := logging.FromContext(ctx)
+
+	selector := labels.Set{serving.ConfigurationLabelKey: config.Name}.AsSelector()
+	revs, err := c.revisionLister.Revisions(config.Namespace).List(selector)
+	if err != nil {
+		return err
+	}
+
+	for _, rev := range revs {
+		if isRevisionStale(ctx, rev, config) {
+			err := c.ServingClientSet.ServingV1alpha1().Revisions(rev.Namespace).Delete(rev.Name, &metav1.DeleteOptions{})
+			if err != nil {
+				logger.Errorf("Failed to delete stale revision: %v", err)
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func isRevisionStale(ctx context.Context, rev *v1alpha1.Revision, config *v1alpha1.Configuration) bool {
+	cfg := configns.FromContext(ctx).RevisionGC
+	logger := logging.FromContext(ctx)
+
+	// maxGen is the maximum generation number we consider for GC
+	maxGen := config.Spec.Generation - cfg.StaleRevisionMinimumGenerations
+
+	if config.Status.LatestReadyRevisionName == rev.Name {
+		return false
+	}
+
+	// Check if rev is within "MinimumGenerations" of latest
+	if gen, err := rev.GetConfigurationGeneration(); err != nil {
+		logger.Errorf("Failed to determine revision configuration generation: %v", err)
+		return false
+	} else if gen > maxGen {
+		return false
+	}
+
+	curTime := time.Now()
+	if rev.ObjectMeta.CreationTimestamp.Add(cfg.StaleRevisionCreateDelay).After(curTime) {
+		// Revision was created sooner than staleRevisionCreateDelay. Ignore it
+		return false
+	}
+
+	lastPin, err := rev.GetLastPinned()
+	if err != nil {
+		if err.(v1alpha1.LastPinnedParseError).Type != v1alpha1.AnnotationParseErrorTypeMissing {
+			logger.Errorf("Failed to determine revision last pinned: %v", err)
+		}
+		return false
+	}
+
+	logger.Infof("Removing stale revision %v with creation time %v and lastPinned time %v.", rev.ObjectMeta.Name, rev.ObjectMeta.CreationTimestamp, lastPin)
+	return lastPin.Add(cfg.StaleRevisionTimeout).Before(curTime)
```

Comment:
The log statement is before the final timestamp check and could lead to confusion